### ISSUE TITLE
idea: default flags

### ISSUE THREAD

**Issue by Luis-Henriquez-Perez** [okamsn/loopy#26]
While I was working on #25. I wanted to check if loopy without the dash destructuring enabled destructured variables in its `with` clause. I could have sworn it did. But anyway, I realized I had already set `loopy-default-destructuring-function` and `loopy-default-accumulation-parsing-function` to the corresponding dash functions and I found it inconvenient to switch back. 

Perhaps instead of asking the user to set these functions we can just have a `loopy-default-flags` custom variable. The variable would contain a list of tags to apply to all loopy loops (unless you override it in particular loop).

Thus enabling certain loopy features would be as simple as:

```elisp
(setq loopy-default-flags '(dash split))
```

**Comment by okamsn** [comment]
> While I was working on #25. I wanted to check if loopy without the dash destructuring enabled destructured variables in its `with` clause. I could have sworn it did. 

It doesn't, but probably should, so I will work on that.

> But anyway, I realized I had already set `loopy-default-destructuring-function` and `loopy-default-accumulation-parsing-function` to the corresponding dash functions and I found it inconvenient to switch back.

For customizing single variables, I use the completing command `customize-set-variable`, which I've bound to `F2 j`. 

> Perhaps instead of asking the user to set these functions we can just have a `loopy-default-flags` custom variable. The variable would contain a list of tags to apply to all loopy loops (unless you override it in particular loop).
> 
> Thus enabling certain loopy features would be as simple as:
> 
> ```emacs-lisp
> (setq loopy-default-flags '(dash split))
> ```

That works for me and would make the code simpler, so I will do this.

**Comment by Luis-Henriquez-Perez** [comment]
Question though: How do you turn off a flag? Say you set `loopy-defualt` to `'(dash)` and want to try out loopy's default destructuring? Maybe you'd do something like `(flag-off dash)` or perhaps `(flag loopy-destructuring)` which would override dash's? Perhaps a good alias for a flag would  be `on`. So you can do `(on dash split)` or `(off dash split)`.

**Comment by okamsn** [comment]
> Question though: How do you turn off a flag? Say you set `loopy-defualt` to
> `'(dash)` and want to try out loopy's default destructuring? Maybe you'd do
> something like `(flag-off dash)` or perhaps `(flag loopy-destructuring)` which
> would override dash's? Perhaps a good alias for a flag would be `on`. So you
> can do `(on dash split)` or `(off dash split)`.

I think (but am not sure) that this could lead to different behavior depending
on the order in which `flag-off` and `flag-on` are handled, if things become
more complicated or the number of options grow.

For returning to the original behavior, it is easy to have a flag like `default`
or some other name.

What are your thoughts on having just the one `flag` macro argument, but having
flags like `no-dash` or `no-split`. The flags could be handled in order, so that
the final flag is applied last. For example, if one had enabled the `dash` flag
by default but wanted to use the default destructuring while also splitting
accumulations, one could do `(flags no-dash split)` or, equivalently, `(flags
default split)`.

Either way (as a note to myself), I think that it makes sense to only allow
disabling for commands listed in `loopy-default-flags`, to avoid accidentally
over-writing values that one doesn't actually want to change.  For example,
`(flags pcase no-dash)` (though I haven't finished the `pcase` flag yet) could
naively setup destructuring with `pcase` and then revert to the default
destructuring.

**Comment by Luis-Henriquez-Perez** [comment]
> that this could lead to different behavior depending
on the order in which flag-off and flag-on are handled

That's a good point.

> What are your thoughts on having just the one flag macro argument, but having
flags like no-dash or no-split

Names like `no-dash` and `no-split` are denoting the absence of a flag as opposed an actual flag, so I don't think they should be legitimate flags by which I mean the set of flags should still be (dash split) not (dash split no-dash no-split). The prefix "no-" then would merely be an indicator of whether the flag is on or off. Perhaps instead of the prefix "no-" in this case we could use a "+" or a "-" like so:

`(flag +dash -split)`

The absence of a "+" or minus would be interpreted as a "+" by default.

`(flag dash -split)`

> The flags could be handled in order, so that
the final flag is applied last. 

We should also address what would happen when incompatible flags are both enabled `(flag +dash +pcase +default)`. Perhaps loopy should issue a warning and this case and as you mentioned accept the last flag. I wonder how we could easily specify the logic of tags.It would be nice to be able to define a tag by specifying its interactions with other tags. For instance +dash implies -pcase.

> For returning to the original behavior, it is easy to have a flag like default
or some other name.

By returning to original behavior, do you mean this `+default` turns off all other tags?

**Comment by okamsn** [comment]
Currently a flag works by settings one or more of 3 variables:

1. `loopy--split-implied-accumulation-results`, which controls the behavior of
   accumulation commands that don't specify a variable
2. `loopy--basic-destructuring-function`, which controls the destructuring of
   iteration commands like `list` or `array` and the destructuring of variable
   listed in the `with` macro arguments
3. `loopy--destructuring-accumulation-parser`, which controls the instructions
   returned from parsing accumulation functions that use destructuring


Changing the behavior is really just a `setq` for variables `let`-bound inside
the macro body.  For example, this is the function that the flag `dash` will
cause to run:

``` emacs-lisp
(defun loopy-dash--flag-setup ()
  "Make this `loopy' loop use Dash destructuring."
  (setq
   loopy--basic-destructuring-function
   #'loopy-dash--destructure-variables
   loopy--destructuring-accumulation-parser
   #'loopy-dash--parse-destructuring-accumulation-command))
```

> > For returning to the original behavior, it is easy to have a flag like
> > `default` or some other name.
> By returning to original behavior, do you mean this `+default` turns off all
> other tags?

By setting the variables to their default values while the code is expanded,
yes.  With how it works now, a flag like `-dash` would just be a limited form of
`+default`, setting only some of the variables to their default values. That
feels redundant, but might be clearer when reading loops that use the flags.

> We should also address what would happen when incompatible flags are both
> enabled `(flag +dash +pcase +default)`. Perhaps loopy should issue a warning
> and this case and as you mentioned accept the last flag. I wonder how we could
> easily specify the logic of tags. It would be nice to be able to define a tag
> by specifying its interactions with other tags. For instance `+dash` implies
> `-pcase`.

For detecting conflicts, an alist would probably be sufficient.

``` emacs-lisp
(defvar list-of-affected-variables
  '((basic-destr . (default pcase))
    (accum-destr . (default pcase))
    (accum-split . (default split))))

;; Add the `dash' symbol, without needing to be aware of the others:
(push 'dash (alist-get 'basic-destr list-of-affected-variables))
(push 'dash (alist-get 'accum-destr list-of-affected-variables))

;; Test if flags in conflict:
(cl-defun conflict-p (local-flags)
  (dolist (conflict list-of-affected-variables)
    (when (> (cl-count-if (lambda (flag) (memq flag (cdr conflict)))
                          local-flags)
             1)
      (cl-return-from conflict-p t)))
  nil)

(conflict-p '(dash pcase)) ; => t
```

However, I think this is maybe complicating the feature more than needed.  It
also wouldn't apply to conflict between flags in `loopy-default-flags` and flags
listed in the `flag` macro argument, which are the most likely to conflict.

As you've noted, `+dash` implies `-pcase`, but what I meant was more that if the
documentation explicitly states that the flags are just applied in order, then
one doesn't absolutely need flags like `-dash` if there is `default` (ignoring
the convenience of `-dash`) and there doesn't need to be a distinction in the
code between the variable `loopy-default-flags` and the `flag` macro argument.
Instead, each flag just keeps setting the variables until the last flag is
applied.

**Comment by Luis-Henriquez-Perez** [comment]
> Instead, each flag just keeps setting the variables until the last flag is
applied.

Ok. That's a good point. Sound good to me. I don't think we need to deal with conflicts.

**Comment by okamsn** [comment]
I've updated how flags behave.

1. There is now a variable `loopy-default-flags`.
2. Flags are applied in the order listed, so that flags in the `flag` macro
   argument override default flags.
3. For convenience, each flag has a `-` version, which will undo the changes the
   flag makes, if those changes are still applied when the `-` version runs.
   - For example, `(flags pcase -dash)` will not stop a user from using `pcase`
     for destructuring. `(flags -dash dash)` leaves Dash destructuring enabled.
   - For clarity, `+dash` is the same as `dash`.
   - The `-` flags just set the variables back to there default values if the
     values are currently what the `+` version set.
4. A `default` flag was added.  There is no `-default` flag.

Do these changes work for you?
